File: solutions/csharp/connect/2/Board.cs
using System.Collections.Generic;
using System.Linq;

internal sealed class Board
{
	private Board() { }
	private Dictionary<HexCoordinates, Cell> _cellByCoordinates;

	internal static Board From(string[] inputString)
		=> inputString == null || inputString.Length == 0
			? throw new global::System.ArgumentException("Input string cannot be null or empty.")
			: new()
			{
				_cellByCoordinates = inputString
					.SelectMany<string, Cell>(Foo)
					.Where<global::Cell>(x => x != null && x.Content != global::BoardContent.Ignored)
					.ToDictionary<global::Cell, global::HexCoordinates, global::Cell>(x => x.Coordinates, x => x)
			};

	private static IEnumerable<Cell> Foo(string rowString, int row)
	{
		bool ignoreSpaces = row % 2 == 0;
		return rowString
			.Where((_, column) =>
				!(ignoreSpaces && column % 2 == 1)
				&& !(!ignoreSpaces && column % 2 == 0)
			)
			.Select((char character, int column) =>
				Cell.From(character, row, column)
			);
	}

	internal Cell Get(HexCoordinates coordinates)
		=> _cellByCoordinates.TryGetValue(coordinates, out Cell cell)
			? cell
			: null;

	internal List<Cell> GetTopEdge()
	{
		int minR = _cellByCoordinates.Keys.Min(c => c.R);
		return _cellByCoordinates.Values
			.Where(cell => cell.Coordinates.R == minR)
			.OrderBy(cell => cell.Coordinates.Q)
			.ToList();
	}

	internal List<Cell> GetBottomEdge()
	{
		int maxR = _cellByCoordinates.Keys.Max(c => c.R);
		return _cellByCoordinates.Values
			.Where(cell => cell.Coordinates.R == maxR)
			.OrderBy(cell => cell.Coordinates.Q)
			.ToList();
	}

	internal List<Cell> GetLeftEdge()
		=> _cellByCoordinates.Values
			.Where(cell =>
				!_cellByCoordinates.ContainsKey(new HexCoordinates(cell.Coordinates.Q - 1, cell.Coordinates.R, cell.Coordinates.S + 1)))
			.GroupBy(cell => cell.Coordinates.R)
			.Select(group => group.OrderBy(cell => cell.Coordinates.Q).First())
			.OrderBy(cell => cell.Coordinates.R)
			.ToList();

	internal List<Cell> GetRightEdge()
		=> _cellByCoordinates.Values
			.Where(cell =>
				!_cellByCoordinates.ContainsKey(new HexCoordinates(cell.Coordinates.Q + 1, cell.Coordinates.R, cell.Coordinates.S - 1)))
			.GroupBy(cell => cell.Coordinates.R)
			.Select(group => group.OrderByDescending(cell => cell.Coordinates.Q).First())
			.OrderBy(cell => cell.Coordinates.R)
			.ToList();

	internal IEnumerable<Cell> GetNeighbors(HexCoordinates coordinates)
		=> new List<HexCoordinates>
		{
				new (coordinates.Q - 1, coordinates.R, coordinates.S + 1),
				new (coordinates.Q - 1, coordinates.R + 1, coordinates.S),
				new (coordinates.Q, coordinates.R - 1, coordinates.S + 1),
				new (coordinates.Q, coordinates.R + 1, coordinates.S - 1),
				new (coordinates.Q + 1, coordinates.R - 1, coordinates.S),
				new (coordinates.Q + 1, coordinates.R, coordinates.S - 1)
		}
		.Select(Get)
		.Where(neighborCell => neighborCell != null);
}

File: solutions/java/connect/1/src/main/java/Board.java
import java.util.*;
import java.util.stream.*;

final class Board {
    private Map<HexCoordinates, Cell> cellByCoordinates;

    private Board() {
    }

    static Board from(String[] inputString) {
        if (inputString == null || inputString.length == 0) {
            throw new IllegalArgumentException("Input string cannot be null or empty.");
        }

        Board board = new Board();
        board.cellByCoordinates = IntStream.range(0, inputString.length)
                                           .mapToObj(row -> Cell.manyFrom(inputString[row], row))
                                           .flatMap(Collection::stream)
                                           .filter(Objects::nonNull)
                                           .filter(cell -> cell.content() != BoardContent.Ignored)
                                           .collect(Collectors.toMap(Cell::coordinates, cell -> cell));
        return board;
    }

    Cell get(HexCoordinates coordinates) {
        return cellByCoordinates.getOrDefault(coordinates, null);
    }

    List<Cell> getTopEdge() {
        int minimumR = cellByCoordinates.keySet()
                                    .stream()
                                    .mapToInt(HexCoordinates::r)
                                    .min()
                                    .orElseThrow();

        return cellByCoordinates.values()
                                .stream()
                                .filter(cell -> cell.coordinates().r() == minimumR)
                                .sorted(Comparator.comparing(cell -> cell.coordinates().q()))
                                .toList();
    }

    List<Cell> getBottomEdge() {
        int maximumR = cellByCoordinates.keySet()
                                    .stream()
                                    .mapToInt(HexCoordinates::r)
                                    .max()
                                    .orElseThrow();

        return cellByCoordinates.values()
                                .stream()
                                .filter(cell -> cell.coordinates().r() == maximumR)
                                .sorted(Comparator.comparing(cell -> cell.coordinates().q()))
                                .toList();
    }

    List<Cell> getLeftEdge() {
        return cellByCoordinates.values()
                                .stream()
                                .filter(cell -> {
                                    HexCoordinates coordinates = cell.coordinates();
                                    HexCoordinates westNeighbor
                                            = new HexCoordinates(
                                                    coordinates.q() - 1,
                                                    coordinates.r(),
                                                    coordinates.s() + 1
                                    );
                                    return !cellByCoordinates.containsKey(westNeighbor);
                                })
                                .collect(Collectors.groupingBy(cell -> cell.coordinates().r()))
                                .values()
                                .stream()
                                .map(group -> group.stream()
                                                   .min(Comparator.comparing(cell -> cell.coordinates().q()))
                                                   .orElseThrow()
                                )
                                .sorted(Comparator.comparing(cell -> cell.coordinates().r()))
                                .toList();
    }

    List<Cell> getRightEdge() {
        return cellByCoordinates.values()
                                .stream()
                                .filter(cell -> {
                                    HexCoordinates coordinates = cell.coordinates();
                                    HexCoordinates eastNeighbor = new HexCoordinates(
                                            coordinates.q() + 1,
                                            coordinates.r(),
                                            coordinates.s() - 1
                                    );

                                    return !cellByCoordinates.containsKey(eastNeighbor);
                                })
                                .collect(Collectors.groupingBy(c -> c.coordinates().r()))
                                .values()
                                .stream()
                                .map(group -> group.stream()
                                                   .max(Comparator.comparing(cell -> cell.coordinates().q()))
                                                   .orElseThrow()
                                )
                                .sorted(Comparator.comparing(cell -> cell.coordinates().r()))
                                .toList();
    }

    Stream<Cell> getNeighbors(HexCoordinates coordinates) {
        List<HexCoordinates> neighbors = List.of(
                new HexCoordinates(coordinates.q() - 1, coordinates.r(), coordinates.s() + 1),
                new HexCoordinates(coordinates.q() - 1, coordinates.r() + 1, coordinates.s()),
                new HexCoordinates(coordinates.q(), coordinates.r() - 1, coordinates.s() + 1),
                new HexCoordinates(coordinates.q(), coordinates.r() + 1, coordinates.s() - 1),
                new HexCoordinates(coordinates.q() + 1, coordinates.r() - 1, coordinates.s()),
                new HexCoordinates(coordinates.q() + 1, coordinates.r(), coordinates.s() - 1)
        );
        return neighbors.stream()
                        .map(this::get)
                        .filter(Objects::nonNull);
    }
}